http/cookies.lua

```diff
@@ -0,0 +1,356 @@
+	cookie.last_access = time or os.time()
+	local domain, path, key = cookie.domain, cookie.path, cookie.key
+	local cookies = self.cookies
+	if cookies[domain] and cookies[domain][path] then
+		local old_cookie = cookies[domain][path][key]
+		if old_cookie then
+			cookie.creation = old_cookie.creation
+		end
+	end
+
+	local old_cookie = self:get(cookie.domain, cookie.path, cookie.key)
+	if old_cookie then
+		self:remove_cookie(old_cookie)
+	end
+	if cookie.persistent then
+		local cookie_exp_time = cookie.expires
+		local inserted = false
+		for i=1, #cookies do
+			-- insert into first spot where cookie expires after
+			if cookies[i].expires < cookie_exp_time then
+				inserted = true
+				table.insert(cookies, i, cookie)
+			end
+		end
+		if not inserted then
+			cookies[#cookies + 1] = cookie
+		end
+	else
+		cookie.expires = math.huge
+		table.insert(cookies, 1, cookie)
+	end
+
+	local by_domain = cookies[domain]
+	if not by_domain then
+		by_domain = {}
+		cookies[domain] = by_domain
+	end
+	local by_path = by_domain[path]
+	if not by_path then
+		by_path = {}
+		by_domain[path] = by_path
+	end
+	by_path[key] = cookie
+end
+
+function cookiejar_methods:get(domain, path, key)
+	local cookies = self.cookies
+	local by_domain = cookies[domain]
+	if not by_domain then
+		return
+	end
+	local by_path = by_domain[path]
+	if not by_path then
+		return
+	end
+	return by_path[key]
+end
+
+local function clear_holes(tbl, n)
+	local start_hole = 0
+	for i=1, n do
+		if tbl[i] and start_hole ~= 0 then
+			tbl[start_hole] = tbl[i]
+			tbl[i] = nil
+			start_hole = start_hole + 1
+		elseif not tbl[i] and start_hole == 0 then
+			start_hole = i
+		end
+	end
+end
+
+function cookiejar_methods:remove_cookie(cookie)
+	local cookies = self.cookies
+	for i=1, #cookies do
+		if cookie == cookies[i] then
+			table.remove(cookies, i)
+			cookies[cookie.domain][cookie.path][cookie.key] = nil
+			return true
+		end
+	end
+	return false
+end
+
+function cookiejar_methods:remove_cookies(cookies)
+	local cookie_hashes = {}
+	for _, key in pairs(cookies) do
+		cookie_hashes[key] = true
+	end
+	local s_cookies = self.cookies
+	local n = #s_cookies
+	for index, value in pairs(s_cookies) do
+		if cookie_hashes[value] then
+			s_cookies[index] = nil
+			local by_domain = s_cookies[value.domain]
+			local by_path = by_domain[value.path]
+			by_path[value.key] = nil
+			if not next(by_path) then
+				by_domain[value.path] = nil
+				if not next(by_domain) then
+					s_cookies[value.domain] = nil
+				end
+			end
+		end
+	end
+
+	clear_holes(s_cookies, n)
+end
+
+local function get_expired(jar, time)
+	time = time or os.time()
+	local cookies = jar.cookies
+	local returned_cookies = {}
+	for i=#cookies, 1, -1 do
+		local cookie = cookies[i]
+		if cookie.expires > time then
+			break
+		end
+		returned_cookies[#returned_cookies + 1] = cookie
+	end
+	return returned_cookies
+end
+
+function cookiejar_methods:remove_expired(time)
+	self:remove_cookies(get_expired(self, time))
```

Comment:
Could inline `:remove_cookies` into here.

Reply:
I'd rather not, given you should still be able to remove cookies from other methods - such as if they're not persistent and a session is closed.